juju/juju
# [JUJU-3969] Allow upgrades without specifying version

The following allows migrating to a new version of juju without explicitly stating the newer major version (3.x -> 4.x). Juju shouldn't have to know about new major versions, so to fix this requires removing the code that was recently added to pass the target controller version to the source pre-checks. In addition to this, I've also refactored the migration precheck context types to not allow target pre-check to explicitly know about model checking. This should prevent any target and source issues.

This is required as 3.1 was earmarked as the last LTS before releasing 4.0. The target controller will then validate where the model is coming from, and ensure it's within its own range before allowing it to migrate.

## Checklist


## QA steps

We require this [PR](https://github.com/juju/juju/pull/15715) one to land before we can fully test the workflow.

To ensure we didn't break any backward compatibility:

```sh
$ sudo snap install juju --channel 3.0/stable
$ juju bootstrap lxd test30
$ juju add-model default
$ juju deploy ubuntu
$ sudo snap remove juju --purge
$ make go-install
$ juju bootstrap lxd test31
$ juju switch controller
$ juju wait-for application controller
$ juju switch test30
$ juju wait-for application ubuntu
$ sleep 10
$ juju migrate default test31
$ juju switch test31
$ sleep 10
$ juju switch default
$ juju wait-for application ubuntu 
```
